### ISSUE TITLE
Adding an example for a 3rd party API service that requires basic authentication

### ISSUE THREAD

**Issue by mitchell-ua** [janwilmake/openapi-mcp-server#1]
While an example of a local API server is good, it will be more helpful to add an example for a 3rd party API service that requires authentication (at least basic login/password).

**Comment by snaggle-ai** [comment]
@mitchell-ua sure! There are soooo many to choose from of course.

Do you have a specific kind of authentication in mind?

I know that OpenAI supports 
- static tokens as either Bearer Auth Tokens or as Custom Headers
- A version of OAuth2 Client Flow

If we flush out a little more detail I can plug it into cursor and implement something

**Comment by mitchell-ua** [comment]
@snaggle-ai Thanks for getting back! My API service supports OAuth2-compatible tokens which are used as Authorization: Bearer. See https://www.oauth.com/oauth2-servers/access-tokens/

**Comment by snaggle-ai** [comment]
No worries. @mitchell-ua how do you get an access token? Is it via an [Authorization Code](https://www.oauth.com/oauth2-servers/access-tokens/authorization-code-request/) flow ?

I think that supporting auth code flow would mean:
 - configuring your auth server with aa redirect to localhost for 
 - providing the mcp server with a client secret as part of setup (could go into the claude developer json but I think it's probably better to put it on the file system somewhere like `~/.openapi-mcp-server` and equivalent for Windows)
 - having the mcp server listen on localhost for a redirect from your auth endpoint
 
There are other options with some auth servers though

**Comment by snaggle-ai** [comment]
@mitchell-ua OK nice and simple!

Would you want to have the user enter the username + password into Claude? Or would you want to put them in a config file or into the claude config json as part of the startup command for the mcp server?

**Comment by mitchell-ua** [comment]
@snaggle-ai I think adding `env` with credentials to `claude_desktop_config.json` is a way to go.

**Comment by snaggle-ai** [comment]
Great, just to confirm the interface, something like this:
```
{
  "mcpServers": {
    "github": {
      "command": "npx",
      "args": ["-y", "openapi-mcp-server", "http://yourdomain.com/openapi.json"],
      "env": {
        "OAUTH_USERNAME": "your_token_here"
        "OAUTH_PASSWORD": "your_token_here"
      }
    }
  }
}
```

**Comment by mitchell-ua** [comment]
Not sure what the key -y does in this case, but I would say:
```
{
  "mcpServers": {
    "third-party-api": {
      "command": "npx",
      "args": ["openapi-mcp-server", "http://yourdomain.com/openapi.json"],
      "env": {
        "OAUTH_USERNAME": "your_username_here",
        "OAUTH_PASSWORD": "your_password_here"
      }
    }
  }
}
```

**Comment by snaggle-ai** [comment]
the `-y` automatically approves any npx prompts that might arise - the official @modelcontextprotocol node examples all use it, I have an update to the README for this project to push to do the same

**Comment by snaggle-ai** [comment]
So, I've investigated this a bit more and the tricky part is that there's no consistent standard structure of request/response that the token/authorization endpoints have to stick to for their request or response payload.

for example some password flow endpoints use http basic auth (no request body at all) and return access token as text, some need a POST request with form-data fields like `username` and `password`, some (the one that you shared @mitchell-ua) need a POST request with json data and the field names vary - the one you shared wants a POST of `{ login:, password: }` but others might use `{ username:, password: }` or `{accountId:, key: }` or basically it could be anything.

*if* we could involve the LLM in the login flow a little it could probably navigate this fairly straightforwardly, although it might be a bit of a PITA user experience.

Alternatively I could encode a bit of a trial-and-error/shotgun approach where the server attempts a load of different approaches and sees if one succeeds, but that will make error reporting very difficult/ confusing.

**Comment by jscastanoc** [comment]
@snaggle-ai Have you gotten any further with this? Seems like involving the LLM would take away the flexibility quite a bit. Probably putting this in the config file would be the best and  depending on the type of authorization, the server manages it accordingly?

For example, one config could look like:

```
{
  "mcpServers": {
    "mydomainapi": {
      "command": "npx",
      "args": ["-y", 
         "openapi-mcp-server", 
        "http://yourdomain.com/openapi.json",
        "--username", "<USERNAME>"
        "--password", "<PASSWORD>"
        "--authtype", "OAUTH"
        "],
    }
  }
}
```

and a different one one
```
{
  "mcpServers": {
    "mydomainapi": {
      "command": "npx",
      "args": ["-y", 
         "openapi-mcp-server", 
        "http://yourdomain.com/openapi.json",
        "--appkey", "key"
        "--authtype", "HTTP"
        "],
    }
  }
}
```

what do you think?

**Comment by snaggle-ai** [comment]
thanks @jscastanoc - I think you're right that this just needs to be user configurable. It needs to be something that can be quite exressive to support a range of cases. I think it has to be namespaced a bit or support some kind of templating.

We have to allow specifying pretty much everything that could go into a fetch call:

url (domain, path, query string)
http method
headers
body
format (json, formdata, xml)

with templating I could skip auto format and just require the provided template is correct for the provided headers.

So I'm thinking of making the open api spec and some arguments available for templating  like 
 - securityScheme.<name>.tokenUrl
 - servers[i].url (from the open api server)
 - args.[name]

example template file using mustache templates (`{{`)
```
{
  url: "{{securityScheme.<name>.tokenUrl}}",
  method: POST,
  headers: {
    "Content-Type": "application/json"
  },
  body: "{ username: "{{ args.username }}", password: "{{ args.password }}" }"
}
```

or for form encoded you could do:
```
{
  url: "{{securityScheme.<name>.tokenUrl}}",
  method: POST,
  headers: {
    "Content-Type": "application/x-www-form-urlencoded"
  },
  body: "login={{ args.username }}&key={{ args.key }}"
}
```

**Comment by Lordinaire** [comment]
Hello, it seems to be difficult to handle all authentification shemes. An alternative could be to provide some configuration to add custom headers (like Authorization/Cookies).

I don't think that authentification must be done by the MCP server.

**Comment by alexanderjeurissen** [comment]
I agree, it's going to be difficult to handle all authentication schemes. To this end I created a simple pull-request to add support for passing the headers used by the HttpClient as a json string through the env variables.

**Comment by snaggle-ai** [comment]
Thanks all, it is a complex area, we've made a start thanks to @alexanderjeurissen.

v.1.1.0 has been released which supports setting custom headers that can be used for some auth cases:

```
{
  "mcpServers": {
    "openapi": {
      "command": "npx",
      "args": ["-y", "openapi-mcp-server@1.1.0", "http://yourdomain.com/openapi.json"],
      "env": {
        "OPENAPI_MCP_HEADERS": "{\"Authorization\": \"Bearer TOKEN\"}"
      }
    }
  }
}
```

**Comment by janwilmake** [comment]
Due to the fact that auth is hard to solve still with MCP, openapi-mcp 2.0 onwards does not include adding the openapi as tools. This MCP server is mainly focused on making it easier to do codegen with complex apis as it can easily find the right endpoints and their spec.

However, it features getting easy navigation through huge OpenAPIs.

For adding support of openapis as tools, please open a discussion to discuss a better solution.
